File: tests/e2e/fixtures/unsigned_integers/mysql/go/runtime_gen_test.go
package main

import (
	"context"
	crand "crypto/rand"
	"database/sql"
	"encoding/hex"
	"math"
	"os"
	"strings"
	"testing"

	gen "e2e-unsigned-integers-go-mysql/gen"

	_ "github.com/go-sql-driver/mysql"
)

func mysqlDSN() string {
	if v := os.Getenv("MYSQL_DSN"); v != "" {
		return v
	}
	return "sqltgen:sqltgen@tcp(localhost:13306)/sqltgen_e2e?parseTime=true"
}

func mysqlAdminDSN() string {
	d := mysqlDSN()
	if idx := strings.LastIndex(d, "/"); idx != -1 {
		rest := d[idx+1:]
		if qIdx := strings.Index(rest, "?"); qIdx != -1 {
			return d[:idx+1] + rest[qIdx:]
		}
		return d[:idx+1]
	}
	return d
}

func randomDBName() string {
	b := make([]byte, 8)
	_, _ = crand.Read(b)
	return "test_" + hex.EncodeToString(b)
}

func setupDB(t *testing.T) (*sql.DB, func()) {
	t.Helper()

	admin, err := sql.Open("mysql", mysqlAdminDSN())
	if err != nil {
		t.Fatal(err)
	}

	dbName := randomDBName()
	if _, err := admin.Exec("CREATE DATABASE " + dbName); err != nil {
		admin.Close()
		t.Fatal(err)
	}

	d := mysqlDSN()
	var testDSN string
	if idx := strings.LastIndex(d, "/"); idx != -1 {
		rest := d[idx+1:]
		if qIdx := strings.Index(rest, "?"); qIdx != -1 {
			testDSN = d[:idx+1] + dbName + rest[qIdx:]
		} else {
			testDSN = d[:idx+1] + dbName
		}
	} else {
		testDSN = d + "/" + dbName
	}

	db, err := sql.Open("mysql", testDSN)
	if err != nil {
		admin.Exec("DROP DATABASE IF EXISTS " + dbName)
		admin.Close()
		t.Fatal(err)
	}

	ddl, err := os.ReadFile("../schema.sql")
	if err != nil {
		db.Close()
		admin.Exec("DROP DATABASE IF EXISTS " + dbName)
		admin.Close()
		t.Fatal(err)
	}
	for _, stmt := range strings.Split(string(ddl), ";") {
		stmt = strings.TrimSpace(stmt)
		if stmt == "" {
			continue
		}
		if _, err := db.Exec(stmt); err != nil {
			db.Close()
			admin.Exec("DROP DATABASE IF EXISTS " + dbName)
			admin.Close()
			t.Fatal(err)
		}
	}

	return db, func() {
		db.Close()
		admin.Exec("DROP DATABASE IF EXISTS " + dbName)
		admin.Close()
	}
}

func TestUnsignedIntegersRoundTripThroughFullRange(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	if err := gen.InsertUnsignedRow(ctx, db, 0, 0, 0, 0, 0); err != nil {
		t.Fatal(err)
	}
	if err := gen.InsertUnsignedRow(ctx, db, 1, 1, 1, 1, 1); err != nil {
		t.Fatal(err)
	}
	// BIGINT UNSIGNED max = 2^64 - 1 = math.MaxUint64.
	if err := gen.InsertUnsignedRow(ctx, db, math.MaxUint8, math.MaxUint16, 16_777_215, math.MaxUint32, math.MaxUint64); err != nil {
		t.Fatal(err)
	}

	rows, err := gen.GetUnsignedRows(ctx, db)
	if err != nil {
		t.Fatal(err)
	}
	if len(rows) != 3 {
		t.Fatalf("expected 3 rows, got %d", len(rows))
	}

	for i, want := range [...]struct {
		u8  uint8
		u16 uint16
		u24 uint32
		u32 uint32
		u64 uint64
	}{
		{0, 0, 0, 0, 0},
		{1, 1, 1, 1, 1},
		{math.MaxUint8, math.MaxUint16, 16_777_215, math.MaxUint32, math.MaxUint64},
	} {
		r := rows[i]
		if r.U8Val != want.u8 || r.U16Val != want.u16 || r.U24Val != want.u24 || r.U32Val != want.u32 || r.U64Val != want.u64 {
			t.Errorf("row %d mismatch: got (%d, %d, %d, %d, %d), want (%d, %d, %d, %d, %d)",
				i, r.U8Val, r.U16Val, r.U24Val, r.U32Val, r.U64Val,
				want.u8, want.u16, want.u24, want.u32, want.u64)
		}
	}

	// The id column itself is BIGINT UNSIGNED.
	if rows[0].Id != 1 || rows[2].Id != 3 {
		t.Errorf("Id round-trip failed: got %d/%d, want 1/3", rows[0].Id, rows[2].Id)
	}
}

File: tests/e2e/fixtures/unsigned_integers/mysql/javascript/runtime_gen.test.js
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { randomBytes } from 'node:crypto';
import mysql from 'mysql2/promise';

import * as queries from './gen/queries/index.js';

const FIXTURES = join(import.meta.dirname, '..');
const MYSQL_HOST = process.env['MYSQL_HOST'] ?? '127.0.0.1';
const MYSQL_PORT = parseInt(process.env['MYSQL_PORT'] ?? '13306', 10);

const ROOT_CONFIG = {
  host: MYSQL_HOST,
  port: MYSQL_PORT,
  user: 'root',
  password: 'sqltgen',
  database: 'sqltgen_e2e',
};

const TEST_CONFIG = {
  host: MYSQL_HOST,
  port: MYSQL_PORT,
  user: 'sqltgen',
  password: 'sqltgen',
  // Required so the driver returns BIGINT UNSIGNED as a JS bigint instead of
  // a string or a lossy Number; without this the round-trip silently truncates.
  supportBigNumbers: true,
  bigNumberStrings: false,
};

async function makeConn() {
  const dbName = 'test_' + randomBytes(16).toString('hex');
  const admin = await mysql.createConnection(ROOT_CONFIG);
  await admin.execute(`CREATE DATABASE \`${dbName}\``);
  await admin.execute(`GRANT ALL ON \`${dbName}\`.* TO 'sqltgen'@'%'`);
  await admin.end();
  const db = await mysql.createConnection({ ...TEST_CONFIG, database: dbName });
  const schemaSql = readFileSync(join(FIXTURES, 'schema.sql'), 'utf8');
  for (const stmt of schemaSql.split(';').map(s => s.trim()).filter(Boolean)) {
    await db.execute(stmt);
  }
  return { db, dbName };
}

async function teardown(db, dbName) {
  await db.end();
  const admin = await mysql.createConnection(ROOT_CONFIG);
  await admin.execute(`DROP DATABASE IF EXISTS \`${dbName}\``);
  await admin.end();
}

describe('UNSIGNED integers', () => {
  it('round-trip through full unsigned range', async () => {
    const { db, dbName } = await makeConn();
    try {
      await queries.insertUnsignedRow(db, 0, 0, 0, 0, 0n);
      await queries.insertUnsignedRow(db, 1, 1, 1, 1, 1n);
      const u64Max = (1n << 64n) - 1n;
      await queries.insertUnsignedRow(db, 255, 65535, 16777215, 4294967295, u64Max);

      const rows = await queries.getUnsignedRows(db);
      assert.equal(rows.length, 3);

      assert.deepEqual([rows[0].u8_val, rows[0].u16_val, rows[0].u24_val, rows[0].u32_val], [0, 0, 0, 0]);
      assert.equal(rows[0].u64_val, 0n);

      assert.deepEqual([rows[1].u8_val, rows[1].u16_val, rows[1].u24_val, rows[1].u32_val], [1, 1, 1, 1]);
      assert.equal(rows[1].u64_val, 1n);

      assert.equal(rows[2].u8_val, 255);
      assert.equal(rows[2].u16_val, 65535);
      assert.equal(rows[2].u24_val, 16777215);
      assert.equal(rows[2].u32_val, 4294967295);
      // The critical correctness gate: 2^64-1 must round-trip without truncation.
      assert.equal(rows[2].u64_val, u64Max);

      assert.equal(rows[0].id, 1n);
      assert.equal(rows[2].id, 3n);
    } finally { await teardown(db, dbName); }
  });
});
